File: android-native/app/src/main/java/com/example/kpkn/domain/performance/PerformanceRangeCalculator.kt
package com.example.kpkn.domain.performance

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.sqrt

data class PerformanceRangeResult(
    val ermMin: Double,
    val ermMax: Double,
    val ermRms: Double,
    val ewmaErm: Double,          // eRM dinámico EWMA con bandwidth capping
    val isCurrentInRange: Boolean,
)

object PerformanceRangeCalculator {

    const val DEFAULT_WINDOW_WEEKS = 8
    private const val OUTLIER_STDDEV_THRESHOLD = 2.0

    // Bandwidth de corrección (Fase 1)
    const val OPTIMISM_CAP_PCT  = 0.07   // +7% techo — una sesión no puede subir más que esto el eRM dinámico
    const val PESSIMISM_FLOOR_PCT = 0.10 // −10% suelo — protege contra crasheos extremos de una sola sesión

    /**
     * Aplica el bandwidth de corrección a [sessionErm] respecto del [previousEwma].
     * Si no hay referencia previa, devuelve [sessionErm] tal cual.
     */
    fun clampSessionErm(sessionErm: Double, previousEwma: Double): Double {
        if (previousEwma <= 0.0) return sessionErm
        val upper = previousEwma * (1.0 + OPTIMISM_CAP_PCT)
        val lower = previousEwma * (1.0 - PESSIMISM_FLOOR_PCT)
        return sessionErm.coerceIn(lower, upper)
    }

    /**
     * Calcula el nuevo eRM dinámico EWMA aplicando bandwidth capping.
     *
     * Fórmula:
     *   clampedSessionErm = clamp(sessionErm, previousEwma ± bandwidth)
     *   eRM_D_nuevo = clampedSessionErm * 0.3 + previousEwma * 0.7
     *
     * @param sessionErm  eRM estimado para la sesión actual.
     * @param previousEwma  Último valor EWMA guardado (0 si no existe).
     */
    fun computeEwmaErm(sessionErm: Double, previousEwma: Double): Double {
        val clamped = clampSessionErm(sessionErm, previousEwma)
        return if (previousEwma <= 0.0) {
            sessionErm   // primer dato: el EWMA arranca en el valor crudo
        } else {
            clamped * 0.3 + previousEwma * 0.7
        }
    }

    fun computeRange(
        snapshots: List<Double>,
        currentErm: Double,
        previousEwma: Double = 0.0,
    ): PerformanceRangeResult {
        val newEwma = computeEwmaErm(currentErm, previousEwma)

        if (snapshots.size < 3) {
            return PerformanceRangeResult(
                ermMin = currentErm * 0.9,
                ermMax = currentErm * 1.1,
                ermRms = currentErm,
                ewmaErm = newEwma,
                isCurrentInRange = true,
            )
        }

        val mean = snapshots.average()
        val variance = snapshots.map { (it - mean).let { d -> d * d } }.average()
        val stddev = sqrt(variance)

        val filtered = if (stddev > 0.0) {
            snapshots.filter { abs(it - mean) <= OUTLIER_STDDEV_THRESHOLD * stddev }
        } else {
            snapshots
        }.ifEmpty { snapshots }

        val rms = sqrt(filtered.map { it * it }.average())

        val size = filtered.size
        val weights = filtered.indices.map { i -> exp(-0.1 * (size - 1 - i)) }
        val weightSum = weights.sum()
        val weightedRms = if (weightSum > 0.0) {
            filtered.indices.sumOf { i -> filtered[i] * weights[i] } / weightSum
        } else {
            rms
        }

        val ermMin = filtered.minOrNull() ?: currentErm * 0.9
        val ermMax = filtered.maxOrNull() ?: currentErm * 1.1

        return PerformanceRangeResult(
            ermMin = ermMin,
            ermMax = ermMax,
            ermRms = weightedRms,
            ewmaErm = newEwma,
            isCurrentInRange = currentErm in ermMin..ermMax,
        )
    }
}

File: ios-native/KPKNFit/KPKNFit/Domain/Performance/PerformanceRangeCalculator.swift
import Foundation

struct PerformanceRangeResult {
    let ermMin: Double
    let ermMax: Double
    let ermRms: Double
    let ewmaErm: Double
    let isCurrentInRange: Bool
}

enum PerformanceRangeCalculator {

    static let defaultWindowWeeks = 8
    private static let outlierStddevThreshold = 2.0

    static let optimismCapPct = 0.07   // +7% ceiling
    static let pessimismFloorPct = 0.10 // -10% floor

    /// Applies bandwidth correction to sessionErm relative to previousEwma.
    static func clampSessionErm(sessionErm: Double, previousEwma: Double) -> Double {
        if previousEwma <= 0.0 { return sessionErm }
        let upper = previousEwma * (1.0 + optimismCapPct)
        let lower = previousEwma * (1.0 - pessimismFloorPct)
        return max(lower, min(sessionErm, upper))
    }

    /// Computes the new EWMA eRM with bandwidth capping.
    static func computeEwmaErm(sessionErm: Double, previousEwma: Double) -> Double {
        let clamped = clampSessionErm(sessionErm: sessionErm, previousEwma: previousEwma)
        if previousEwma <= 0.0 {
            return sessionErm
        } else {
            return clamped * 0.3 + previousEwma * 0.7
        }
    }

    static func computeRange(
        snapshots: [Double],
        currentErm: Double,
        previousEwma: Double = 0.0
    ) -> PerformanceRangeResult {
        let newEwma = computeEwmaErm(sessionErm: currentErm, previousEwma: previousEwma)

        if snapshots.count < 3 {
            return PerformanceRangeResult(
                ermMin: currentErm * 0.9,
                ermMax: currentErm * 1.1,
                ermRms: currentErm,
                ewmaErm: newEwma,
                isCurrentInRange: true
            )
        }

        let mean = snapshots.reduce(0, +) / Double(snapshots.count)
        let variance = snapshots.map { ($0 - mean) * ($0 - mean) }.reduce(0, +) / Double(snapshots.count)
        let stddev = sqrt(variance)

        let filtered: [Double]
        if stddev > 0.0 {
            filtered = snapshots.filter { abs($0 - mean) <= outlierStddevThreshold * stddev }
        } else {
            filtered = snapshots
        }
        let safeFiltered = filtered.isEmpty ? snapshots : filtered

        let rms = sqrt(safeFiltered.map { $0 * $0 }.reduce(0, +) / Double(safeFiltered.count))

        let size = safeFiltered.count
        let weights = (0..<size).map { i in exp(-0.1 * Double(size - 1 - i)) }
        let weightSum = weights.reduce(0, +)
        let weightedRms: Double
        if weightSum > 0.0 {
            weightedRms = (0..<size).reduce(0.0) { $0 + safeFiltered[$1] * weights[$1] } / weightSum
        } else {
            weightedRms = rms
        }

        let ermMin = safeFiltered.min() ?? (currentErm * 0.9)
        let ermMax = safeFiltered.max() ?? (currentErm * 1.1)

        return PerformanceRangeResult(
            ermMin: ermMin,
            ermMax: ermMax,
            ermRms: weightedRms,
            ewmaErm: newEwma,
            isCurrentInRange: currentErm >= ermMin && currentErm <= ermMax
        )
    }
}
